packages/apollo-server-core/src/utils/pluginTestHarness.ts

```diff
@@ -0,0 +1,145 @@
+import {
+  WithRequired,
+  GraphQLRequest,
+  GraphQLRequestContextExecutionDidStart,
+  GraphQLResponse,
+  ValueOrPromise,
+  GraphQLRequestContextWillSendResponse,
+  GraphQLRequestContext,
+} from 'apollo-server-types';
+import { GraphQLSchema, GraphQLObjectType, GraphQLString } from 'graphql/type';
+import { CacheHint } from 'apollo-cache-control';
+import {
+  enablePluginsForSchemaResolvers,
+  symbolRequestListenerDispatcher,
+} from '../requestPipelineAPI';
+import { ApolloServerPlugin } from 'apollo-server-plugin-base';
+import { InMemoryLRUCache } from 'apollo-server-caching';
+import { Dispatcher } from './dispatcher';
+
+// This test harness guarantees the presence of `query`.
+type IPluginTestHarnessGraphqlRequest = WithRequired<GraphQLRequest, 'query'>;
+type IPluginTestHarnessExecutionDidStart<TContext> =
+  GraphQLRequestContextExecutionDidStart<TContext> & {
+    request: IPluginTestHarnessGraphqlRequest,
+  };
+
+export default async function pluginTestHarness<TContext>({
+  pluginInstance,
+  schema,
+  graphqlRequest,
+  overallCachePolicy,
+  executor,
+  context = Object.create(null)
+}: {
+  /**
+   * An instance of the plugin to test.
+   */
+  pluginInstance: ApolloServerPlugin<TContext>,
+
+  /**
+   * The optional schema that will be received by the executor.  If not
+   * specified, a simple default schema will be created.  In either case,
+   * the schema will be mutated by wrapping the resolvers with the
+   * `willResolveField` instrumentation that will allow it to respond to
+   * that lifecycle hook's implementations plugins.
+   */
+  schema?: GraphQLSchema;
+
+  /**
+   * The `GraphQLRequest` which will be received by the `executor`.  The
+   * `query` is required, and this doesn't support anything more exotic,
+   * like automated persisted queries (APQ).
+   */
+  graphqlRequest: IPluginTestHarnessGraphqlRequest;
+
+  /**
+   * Overall cache control policy.
+   */
+  overallCachePolicy?: Required<CacheHint>;
+
+  /**
+   * This method will be executed to retrieve the response.
+   */
+  executor: (
+    requestContext: IPluginTestHarnessExecutionDidStart<TContext>,
+  ) => ValueOrPromise<GraphQLResponse>;
+
+  /**
+   * (optional) To provide a user context, if necessary.
+   */
+  context?: TContext;
+}): Promise<GraphQLRequestContextWillSendResponse<TContext>> {
+
+  if (!schema) {
+    schema = new GraphQLSchema({
+      query: new GraphQLObjectType({
+        name: 'RootQueryType',
+        fields: {
+          hello: {
+            type: GraphQLString,
+            resolve() {
+              return 'hello world';
+            }
+          }
+        }
+      })
+    });
+  }
+
+  enablePluginsForSchemaResolvers(schema);
+
+  const requestContext: GraphQLRequestContext<TContext> = {
+    logger: console,
```

Comment:
How do you feel about allowing a `logger` to be passed in to the harness? I found I quite liked this pattern to clean up logs during test runs and to simplify writing expectations around mocking.

https://github.com/apollographql/apollo-server/blob/fcbff5bc358f938f169d5b998880cad8d171cf43/packages/apollo-gateway/src/__tests__/integration/networkRequests.test.ts#L88-L98

Reply:
I feel great! This is a great suggestion.  I suspect we'll want to make this more of a first-class experience in our Jest test setup eventually by making it more globally available? (As a helper or otherwise.)  Anyhow, let's certainly cross that bridge the next time we're presented with an opportunity to expand on this pattern.

Concretely right now, per your suggestion, I've added `logger` in https://github.com/apollographql/apollo-server/pull/3990/commits/dae59a5bbee72c4f7de8eb71f8fa4f0f3afe1f2f.  Thanks!